File: grid.py
from typing import Any, Dict, List, Optional, Tuple

import consts
from consts import (
  BLANK, Dir, opposite,
)

class GridDef:
  def __init__(self, w: int, h: int, mult_squares: Dict[int, str]):
    self.w = w
    self.h = h
    self.mult_squares = mult_squares


class CPos:
  def __init__(self, grid: GridDef, i: int):
    self.grid = grid
    self.i = i
    self.x = i % grid.w
    self.y = i // grid.w

  def __str__(self) -> str:
    return f"({self.x},{self.y})"


  # WARNING: this method assumes the caller knows
  # what they are doing in that the requested
  # traversal will not go off the board
  def traverse(self, ncells: int, dirn: Dir): # -> CPos
    match dirn:
      case Dir.UP:
        return CPos(self.grid, self.i - (self.grid.w * ncells))
      case Dir.DOWN:
        return CPos(self.grid, self.i + (self.grid.w * ncells))
      case Dir.LEFT:
        return CPos(self.grid, self.i - ncells)
      case Dir.RIGHT:
        return CPos(self.grid, self.i + ncells)
      case _:
        raise Exception(f"unrecognized direction: {dirn}")

class Cell:
  def __init__(self, grid, pos: CPos, ctype: str, value: Optional[str] = None):
    self.grid = grid
    self.pos: CPos = pos
    self.ctype: str = ctype
    self.value: Optional[str] = value
    self.added: bool = False

  def __str__(self) -> str:
    val: str = (
      self.value if self.value is not None
      else self.ctype if self.ctype != BLANK else "''"
    )
    return f"{self.pos}={val}"

  def letter_mult(self) -> int:
    match self.ctype:
      case consts.DL:
        return 2
      case consts.TL:
        return 3
      case _:
        return 1

  def word_mult(self) -> int:
    match self.ctype:
      case consts.DW:
        return 2
      case consts.TW:
        return 3
      case _:
        return 1

  def is_middle_cell(self) -> bool:
    return self.pos.x == self.grid.w // 2 and self.pos.y == self.grid.h // 2

  def is_anchored(self):
    up: Optional[Cell] = self.grid.up_from(self)
    down: Optional[Cell] = self.grid.down_from(self)
    left: Optional[Cell] = self.grid.left_from(self)
    right: Optional[Cell] = self.grid.right_from(self)
    return (
      (up is not None and up.value is not None and not up.added)
      or (down is not None and down.value is not None and not down.added)
      or (left is not None and left.value is not None and not left.added)
      or (right is not None and right.value is not None and not right.added)
    )


class Grid(GridDef):
  def __init__(self, width: int, height: int, mult_squares: List[Dict[int, str]], init=True):
    GridDef.__init__(self, width, height, mult_squares)
    if init:
      self.size = self.w * self.h
      self.cells: List[Cell] = []
      for nrow in range(self.h):
        srcrow: int = self.h - 1 - nrow if nrow >= len(mult_squares) else nrow
        row: Dict[int, str] = mult_squares[srcrow]
        for ncol in range(self.w):
          pos: CPos = CPos(self, nrow * self.w + ncol)
          ctype = row[ncol] if ncol in row else BLANK
          self.cells.append(Cell(self, pos, ctype))


  def clone(self) -> Any:
    newgrid: Grid = Grid(self.w, self.h, self.mult_squares, False)
    newgrid.size = self.size
    newgrid.cells = []
    for cell in self.cells:
      newcell: Cell = Cell(newgrid, cell.pos, cell.ctype, cell.value)
      newcell.added = cell.added
      newgrid.cells.append(newcell)
    return newgrid


  def show(self) -> str:
    showstr: str = "   "
    for i in range(self.w):
      showstr += f"{i} ".rjust(3, ' ')
    showstr += "\n"
    for y in range(self.h):
      showstr += f"{y} ".rjust(3, ' ')
      for x in range(self.w):
        cell: Cell = self.at(x, y)
        if cell.value is not None:
          esc: int = 32 if cell.added else 35
          showstr += f" [1;{esc}m{cell.value}[m "
        else:
          showstr += f" {self.at(x, y).ctype}"
      showstr += "\n"

    return showstr

  def apply(self, word: str, startpos: CPos, dirn: Dir) -> None:
    pos: CPos = startpos
    for i in range(len(word)):
      cell: Cell = self.cells[pos.i] 
      if cell.value is None:
        cell.value = word[i]
        cell.added = True
      elif cell.value != word[i]:
        raise Exception(
          f"failure applying word {word}: found existing letter {cell.value} "
          f"at ({pos.x},{pos.y}) instead of {word[i]}"
        )
      pos = pos.traverse(1, dirn)

  def at(self, x: int, y: int) -> Cell:
    return self.cells[self.w * y + x]

  def next_cell(self, cell: Cell, dirn: Dir) -> Optional[Cell]:
    match dirn:
      case Dir.UP:
        return self.up_from(cell)
      case Dir.DOWN:
        return self.down_from(cell)
      case Dir.LEFT:
        return self.left_from(cell)
      case Dir.RIGHT:
        return self.right_from(cell)
    raise Exception(f"unrecognized direction: {dirn}")
        

  def left_from(self, cell: Cell) -> Optional[Cell]:
    return self.cells[cell.pos.i - 1] if cell.pos.x > 0 else None

  def right_from(self, cell: Cell) -> Optional[Cell]:
    return self.cells[cell.pos.i + 1] if cell.pos.x < self.w - 1 else None

  def up_from(self, cell: Cell) -> Optional[Cell]:
    return self.cells[cell.pos.i - self.w] if cell.pos.y > 0 else None

  def down_from(self, cell: Cell) -> Optional[Cell]:
    return self.cells[cell.pos.i + self.w] if cell.pos.y < self.h - 1 else None

  def upstr(self, cell: Cell) -> Tuple[str, Optional[Cell]]:
    strpath, start = self.string(cell, Dir.UP)
    return strpath[::-1], start

  def downstr(self, cell: Cell) -> Tuple[str, Optional[Cell]]:
    return self.string(cell, Dir.DOWN)

  def leftstr(self, cell: Cell) -> Tuple[str, Optional[Cell]]:
    strpath, start = self.string(cell, Dir.LEFT)
    return strpath[::-1], start

  def rightstr(self, cell: Cell) -> Tuple[str, Optional[Cell]]:
    return self.string(cell, Dir.RIGHT)

  # string of letters starting from but not including
  # the passed-in cell in direction dirn, up to a blank
  # cell or the end of the board
  def string(self, cell: Cell, dirn: Dir) -> Tuple[str, Optional[Cell]]:
    string: str = ""
    end: Optional[Cell] = None
    nxt: Optional[Cell] = self.next_cell(cell, dirn)
    while (nxt is not None and nxt.value is not None):
      string += nxt.value
      end = nxt
      nxt = self.next_cell(nxt, dirn)

    return string, end


def grid_from_file(fname: str, w: int, h: int, mult_squares: Dict[int, str]) -> Grid:
  grid: Grid = Grid(w, h, mult_squares)
  with open(fname, "r") as file:
    line: int = 0
    while line < grid.h:
      row: str = file.readline().strip()
      for i, s in enumerate(row):
        if s != "-" and s != " ":
          grid.at(i, line).value = s
      line += 1
  return grid


def grid_add_letter(grid: Grid, pos: CPos, letter: str) -> None:
  grid.cells[pos.i].value = letter
  grid.cells[pos.i].added = True

File: go/grid.go
package main

import (
	"errors"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type CPos struct {
	gridw, i, x, y int
}

func NewCPos(gridw, i int) CPos {
	return CPos{gridw, i, i % gridw, i / gridw}
}

func (pos CPos) String() string {
	return fmt.Sprintf("(%d,%d)", pos.x, pos.y)
}

// WARNING: this method assumes the caller knows
// what they are doing in that the requested
// traversal will not go off the board
func (pos CPos) Traverse(ncells int, dirn Dir) CPos {
	switch dirn {
	case UP:
		return NewCPos(pos.gridw, pos.i-(pos.gridw*ncells))
	case DOWN:
		return NewCPos(pos.gridw, pos.i+(pos.gridw*ncells))
	case LEFT:
		return NewCPos(pos.gridw, pos.i-ncells)
	case RIGHT:
		return NewCPos(pos.gridw, pos.i+ncells)
	default:
		return pos
	}
}

type Cell struct {
	pos   CPos
	ctype string
	value string
	added bool
}

func (cell Cell) String() string {
	val := cell.value
	if val == "" {
		switch cell.ctype {
		case BLANK:
			val = ""
		default:
			val = cell.ctype
		}
	}
	return fmt.Sprintf("%s:%s", cell.pos, val)
}

func (cell Cell) LetterMult() int {
	switch cell.ctype {
	case DL:
		return 2
	case TL:
		return 3
	default:
		return 1
	}
}

func (cell Cell) WordMult() int {
	switch cell.ctype {
	case DW:
		return 2
	case TW:
		return 3
	default:
		return 1
	}
}

type Grid struct {
	w, h  int
	tiles []RowSpec
	cells []Cell
}

func NewGrid(board BoardSpec) *Grid {
	newgrid := &Grid{board.w, board.h, board.tiles, make([]Cell, board.w*board.h)}
	for nrow := 0; nrow < newgrid.h; nrow++ {
		var srcrow int = nrow
		if nrow >= len(newgrid.tiles) {
			srcrow = newgrid.h - 1 - nrow
		}
		var row RowSpec = newgrid.tiles[srcrow]
		for ncol := 0; ncol < newgrid.w; ncol++ {
			pos := NewCPos(newgrid.w, nrow*newgrid.w+ncol)
			ctype := row[ncol]
			if ctype == "" {
				ctype = BLANK
			}
			newgrid.cells[pos.i] = Cell{pos, ctype, "", false}
		}
	}

	if newgrid.board_has_unknowns(board) {
		print("[33;mThis board has unknowns![m")
		print("\n", newgrid.show())
		Input("\nHit return to continue")
	}

	return newgrid
}

func (g Grid) board_has_unknowns(board BoardSpec) bool {
	for _, row := range board.tiles {
		for _, ctype := range row {
			if ctype == UK {
				return true
			}
		}
	}
	return false
}

func (g Grid) at(x, y int) *Cell {
	return &g.cells[g.w*y+x]
}

func (g Grid) Serialize() string {
	str := fmt.Sprintf("%d:%d", g.w, g.h)
	for _, cell := range g.cells {
		if cell.value != "" {
			str += cell.value
		} else {
			switch cell.ctype {
			case BLANK:
				str += " "
			case DL:
				str += "-"
			case TL:
				str += "_"
			case DW:
				str += "="
			case TW:
				str += "+"
			}
		}
	}
	return str
}

func deserialize(str string) *Grid {
	var split []string = strings.Split(str, ":")
	sw, sh, data := split[0], split[1], split[2]
	w, err := strconv.Atoi(sw)
	h, err2 := strconv.Atoi(sh)
	if err != nil || err2 != nil {
		fmt.Fprintf(os.Stderr, "bad width or height string: w=%s h=%s\n", sw, sh)
		return nil
	}
	grid := NewGrid(BoardSpec{w, h, []RowSpec{}})
	for i, ch := range data {
		pos := NewCPos(grid.w, i)
		switch ch {
		case ' ':
			grid.cells[i] = Cell{pos, BLANK, "", false}
		case '-':
			grid.cells[i] = Cell{pos, DL, "", false}
		case '_':
			grid.cells[i] = Cell{pos, TL, "", false}
		case '=':
			grid.cells[i] = Cell{pos, DW, "", false}
		case '+':
			grid.cells[i] = Cell{pos, TW, "", false}
		default:
			grid.cells[i] = Cell{pos, BLANK, string(ch), false}
		}
	}
	return grid
}

func (g Grid) clone() Grid {
	newgrid := Grid{g.w, g.h, g.tiles, make([]Cell, g.w*g.h)}
	for i, cell := range g.cells {
		newgrid.cells[i] = Cell{cell.pos, cell.ctype, cell.value, cell.added}
	}
	return newgrid
}

func (g Grid) show() string {
	showstr := "   "
	for i := 0; i < g.w; i++ {
		showstr += fmt.Sprintf("%2s ", strconv.Itoa(i))
	}
	showstr += "\n"
	for y := 0; y < g.h; y++ {
		showstr += fmt.Sprintf("%2s ", strconv.Itoa(y))
		for x := 0; x < g.w; x++ {
			cell := g.at(x, y)
			if cell.value != "" {
				esc := 34
				if cell.added {
					esc = 32
				}
				showstr += fmt.Sprintf(" [1;%dm%s[m ", esc, cell.value)
			} else {
				showstr += fmt.Sprintf(" %s", g.at(x, y).ctype)
			}
		}
		showstr += "\n"
	}

	return showstr
}

// color the found word based on which letters are being added and which
// were already there
func (g Grid) ColorWord(word string, startpos CPos, dirn Dir) (string, error) {
	pos := startpos
	inEsc := false
	coloredStr := ""
	for _, ch := range word {
		cell := g.at(pos.x, pos.y)
		var letter2apply string = string(ch)
		if cell.value == "" {
			if !inEsc {
				coloredStr += "\x1b[1;32m"
				inEsc = true
			}
			coloredStr += letter2apply

		} else if cell.value == letter2apply {
			if inEsc {
				coloredStr += "\x1b[m"
				inEsc = false
			}
			coloredStr += letter2apply

		} else {
			msg := fmt.Sprintf("could not apply %s at %s dir %s to:\n%s",
				word, startpos, dirn, g.show())
			msg += fmt.Sprintf("failure applying word %s: found existing letter %s "+
				"at (%d,%d) instead of '%s'",
				word, cell.value, pos.x, pos.y, letter2apply)
			fmt.Println(msg)
			return "", errors.New(msg)
		}
		pos = pos.Traverse(1, dirn)
	}
	if inEsc {
		coloredStr += "\x1b[m"
	}

	return coloredStr, nil
}

func (g *Grid) apply(word string, startpos CPos, dirn Dir) (*Grid, error) {
	grid := g.clone()
	pos := startpos
	for _, ch := range word {
		cell := grid.at(pos.x, pos.y)
		var letter2apply string = string(ch)
		if cell.value == "" {
			cell.value = letter2apply
			cell.added = true
		} else if cell.value != letter2apply {
			msg := fmt.Sprintf("could not apply %s at %s dir %s to:\n%s",
				word, startpos, dirn, g.show())
			msg += fmt.Sprintf("failure applying word %s: found existing letter %s "+
				"at (%d,%d) instead of '%s'",
				word, cell.value, pos.x, pos.y, letter2apply)
			fmt.Println(msg)
			return nil, errors.New(msg)
		}
		pos = pos.Traverse(1, dirn)
	}

	return &grid, nil
}

func (g Grid) next_cell(cell Cell, dirn Dir) *Cell {
	var next *Cell
	switch dirn {
	case UP:
		next = g.up_from(cell)
	case DOWN:
		next = g.down_from(cell)
	case LEFT:
		next = g.left_from(cell)
	case RIGHT:
		next = g.right_from(cell)
	default:
		panic(fmt.Sprintf("unrecognized direction: %s", dirn))
	}
	return next
}

func (g Grid) left_from(cell Cell) *Cell {
	if cell.pos.x == 0 {
		return nil
	}
	return &g.cells[cell.pos.i-1]
}

func (g Grid) right_from(cell Cell) *Cell {
	if cell.pos.x >= g.w-1 {
		return nil
	}
	return &g.cells[cell.pos.i+1]
}

func (g Grid) up_from(cell Cell) *Cell {
	if cell.pos.y == 0 {
		return nil
	}
	return &g.cells[cell.pos.i-g.w]
}

func (g Grid) down_from(cell Cell) *Cell {
	if cell.pos.y >= g.h-1 {
		return nil
	}
	if cell.pos.i+g.w >= len(g.cells) {
		fmt.Printf("down_from failing: from cell=%v", cell)
	}
	return &g.cells[cell.pos.i+g.w]
}

func (g Grid) StrUpFromCell(cell Cell) (str string, end_cell *Cell) {
	strpath, start := g.StrFromCell(cell, UP)
	return Reverse(strpath), start
}

func (g Grid) StrDownFromCell(cell Cell) (str string, end_cell *Cell) {
	return g.StrFromCell(cell, DOWN)
}

func (g Grid) StrLeftFromCell(cell Cell) (str string, end_cell *Cell) {
	strpath, start := g.StrFromCell(cell, LEFT)
	return Reverse(strpath), start
}

func (g Grid) StrRightFromCell(cell Cell) (str string, end_cell *Cell) {
	return g.StrFromCell(cell, RIGHT)
}

// string of all letters between the passed-in cell in direction
// dirn, up to the first blank cell or edge of the board.
func (g Grid) StrFromCell(cell Cell, dirn Dir) (str string, end_cell *Cell) {
	str = ""
	end_cell = nil
	var nxt *Cell = g.next_cell(cell, dirn)
	for nxt != nil && nxt.value != "" {
		str += nxt.value
		end_cell = nxt
		nxt = g.next_cell(*nxt, dirn)
	}
	return
}

func (g Grid) IsMiddleCell(cell Cell) bool {
	return cell.pos.x == g.w/2 && cell.pos.y == g.h/2
}

func (g Grid) IsAnchored(cell Cell) bool {
	var up *Cell = g.up_from(cell)
	var down *Cell = g.down_from(cell)
	var left *Cell = g.left_from(cell)
	var right *Cell = g.right_from(cell)
	return ((up != nil && up.value != "" && !up.added) ||
		(down != nil && down.value != "" && !down.added) ||
		(left != nil && left.value != "" && !left.added) ||
		(right != nil && right.value != "" && !right.added))
}

func GridFromFile(fname string, board BoardSpec) *Grid {
	grid := NewGrid(board)

	data, err := os.ReadFile(fname)
	if err != nil {
		fmt.Fprintf(os.Stderr, "failed to read grid state from %s\n", fname)
	}
	var rows []string = strings.Split(string(data), "\n")
	var line int = 0
	for line < grid.h {
		var row string = strings.Trim(rows[line], " \r\n")
		for i, ch := range row {
			if ch != '-' && ch != ' ' {
				grid.at(i, line).value = string(ch)
			}
		}
		line += 1
	}
	return grid
}
